types/mocha/index.d.ts

```diff
@@ -1307,9 +1339,7 @@ declare namespace Mocha {
      *
      * @see https://mochajs.org/api/module-Context.html#~Context
      */
-    class Context {
-        private _runnable;
-
+    class BaseContext {
```

Comment:
Does this class actually exist at runtime? This will be exported so implies it can be new'd.

Also, the comment above is attached to the wrong declaration.

Reply:
No, it does not exist.  But this doesn't work with BaseContext as an interface.

To have a strongly-typed extension of a Mocha Context, a base class definition must exist without the index signature, so that types in the extension can have strong type enforcement.  Extending `Context` with its index signature means a field-name typo on the subclass is ignored.

Making BaseContext an interface (no implied constructor) doesn't work at all.  The generic calls that reference it say `T extends BaseContext`, but `T implements BaseContext` doesn't work.

I've fixed the placement of the comment you mentioned, and expanded comments on both BaseContext and Context to clarify their purpose.

I've also added a protected constructor to the definition of BaseContext so that it can't be constructed directly, and I added a new test to verify that.

Reply:
`extends` in a type parameter is not the same as `extends` in a class declaration. In a type parameter it defines a constraint and doesn't care if that constraint is a class or interface.

Reply:
Yeah, I don't know what you mean here. You can just write:

```ts
interface BaseContext {} 

class Context implements BaseContext {}

beforeAll<T extends BaseContext>(fn?: Func<T>): this;
```

Reply:
@jakebailey, I actually tried that (interface BaseContext, class Context implements BaseContext).  But if we do that, then we have to duplicate every member of that interface in Context, or else we get:

```
compile error: 
Class 'Context' incorrectly implements interface 'BaseContext'.
Type 'Context' is missing the following properties from type 'BaseContext': ...
```

I think it's preferable not to duplicate them, because future methods could be accidentally added to Context without adding them also to BaseContext.  By lying about BaseContext being a class, we can avoid repeating ourselves and avoid letting the definitions get out of sync in the future.

Thoughts?

Reply:
Where did you test that? IIRC that kind of error should not appear in a declaration file.

![image](https://github.com/user-attachments/assets/472c9af9-62be-410e-8e04-ac5ab7b6be51)

Reply:
Agh, I had `skipLibCheck` on.

If we need to lie about it, then we need to make sure that the base class is unexported. Otherwise, the library is lying about there being a runtime value there when there is not.

But, I'm not sure it's such a bad thing to have to declare things in the interface? Can't a comment be left?

Reply:
(given mocha's new maintainership, I do also wonder if this will long-term come back to bite us, given maybe it will be converted to TS itself and make these hacks go away, but maybe not)

Reply:
> But, I'm not sure it's such a bad thing to have to declare things in the interface? Can't a comment be left?

Sorry, could you be more explicit?  Are you suggesting we duplicate every method in BaseContext and Context and just leave a comment that people should populate/update both?

If that's what you mean, I don't think one comment would suffice.  If I were adding a missing method without knowing the structure, I would go straight to the bottom of Context and add it.  To know about this, I would probably need to see a comment on _every_ method mentioning its duplicate.  That feels heavy to me.

Ultimately, I want to see this merged, and I **am** willing to compromise on these things if y'all feel strongly.

Reply:
> (given mocha's new maintainership, I do also wonder if this will long-term come back to bite us, given maybe it will be converted to TS itself and make these hacks go away, but maybe not)

If Mocha goes to TS, then DefinitelyTyped won't need to have separate types, right?  If that happens, I can argue for a better structure upstream that accommodates strong types, maybe without an index-signature in the first place.  While Mocha is written in JS, though, the index-signature is implied by the nature of JavaScript.

Reply:
> If Mocha goes to TS, then DefinitelyTyped won't need to have separate types, right?

Mainly I just meant that while we can "lie" here about there being a class, a real project in TS can't do that.